langwatch/src/pages/api/triggers.ts

```diff
@@ -0,0 +1,114 @@
+import { type NextApiRequest, type NextApiResponse } from "next";
+
+import { type Trigger, TriggerAction } from "@prisma/client";
+import { getAllForProject } from "~/server/api/routers/traces";
+import { prisma } from "../../server/db";
+
+import { sendTriggerEmail } from "~/server/mailer/triggerEmail";
+import { getLatestUpdatedAt } from "./utils";
+
+interface ActionParams {
+  members?: string[] | null;
+  dataset?: string | null;
+}
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  if (req.method !== "POST") {
+    return res.status(405).end(); // Only accept POST requests
+  }
+
+  const triggers = await prisma.trigger.findMany({
+    where: {
+      active: true,
+    },
+  });
+
+  const results = [];
+
+  for (const trigger of triggers) {
+    const traces = await getTracesForAlert(trigger);
+    results.push(traces);
+  }
+
+  return res.status(200).json(results);
+}
+
+const getTracesForAlert = async (trigger: Trigger) => {
+  const {
+    id: alertId,
+    projectId,
+    filters,
+    lastRunAt,
+    action,
+    actionParams,
+    name,
+  } = trigger;
+
+  const input = {
+    projectId,
+    filters,
+    updatedAt: lastRunAt,
+  };
+
+  // @ts-ignore
+  const traces = await getAllForProject({}, input);
+
+  if (traces.groups.length > 0) {
+    const triggerData = traces.groups.flatMap((group) =>
+      group.map((trace) => ({
+        input: trace.input?.value,
+        output: trace.output?.value ?? "",
+        traceId: trace.trace_id,
+      }))
+    );
+
+    const project = await prisma.project.findFirst({
+      where: {
+        id: input.projectId,
+      },
+    });
```

Comment:
then you could avoid the n+1 query issue of re-querying each project for each trigger, you could send it together as a parameter